src/ivector/xvector-test.cc

```diff
@@ -0,0 +1,299 @@
+  }
+
+  // Compare the b derivative calculated above with a numerical
+  // approximation.
+  BaseFloat b_p = b + perturb_delta;
+  BaseFloat b_n = b - perturb_delta;
+  CuMatrix<BaseFloat> deriv_xvector_tmp(num_rows, xvector_dim, kSetZero);
+  CuVector<BaseFloat> deriv_S_and_b_tmp(S_dim + 1, kSetZero);
+  BaseFloat tot_objf_p;
+  BaseFloat tot_objf_n;
+  ComputeXvectorObjfAndDeriv(xvector_pairs, S, b_p, &deriv_xvector_tmp,
+    &deriv_S_and_b_tmp, &tot_objf_p, &tot_weight);
+  ComputeXvectorObjfAndDeriv(xvector_pairs, S, b_n, &deriv_xvector_tmp,
+    &deriv_S_and_b_tmp, &tot_objf_n, &tot_weight);
+  BaseFloat delta = (tot_objf_p  - tot_objf_n) * 1.0 / (2.0 * perturb_delta);
+  l2_b = pow(deriv_S_and_b(S_dim) - delta, 2);
+  KALDI_ASSERT(l2_xvector < 1.0e-03);
+  KALDI_ASSERT(l2_S <  1.0e-03);
+  KALDI_ASSERT(l2_b < 1.0e-03);
+  return true;
+}
+
+bool TestXvectorComputeObjf() {
+  int32 xvector_dim = RandInt(4, 30),
+        num_rows = 2 * RandInt(2, 10); // The number of rows must be even
+                                       // and greater than 2.
+  CuSpMatrix<BaseFloat> S(xvector_dim);
+  S.SetRandn();
+  // Necessary to keep the similarity scores from getting too large or small.
+  S.Scale(1.0e-01);
+  BaseFloat b = RandInt(-200, 200) / 10.0,
+            tot_weight,
+            tot_weight_test,
+            tot_objf,
+            tot_objf_test;
+  int32 S_dim = S.NumCols() * (S.NumCols() + 1) / 2;
+  CuMatrix<BaseFloat> xvector_pairs(num_rows, xvector_dim, kSetZero),
+                      deriv_xvector(num_rows, xvector_dim, kSetZero),
+                      deriv_xvector_test(num_rows, xvector_dim, kSetZero);
+  CuVector<BaseFloat> deriv_S_and_b(S_dim + 1, kSetZero),
+                      deriv_S_and_b_test(S_dim + 1, kSetZero);
+  xvector_pairs.SetRandn();
+
+  ComputeXvectorObjfAndDeriv(xvector_pairs, S, b, &deriv_xvector,
+    &deriv_S_and_b, &tot_objf, &tot_weight);
+  TestComputeXvectorObjfAndDeriv(xvector_pairs, S, b, &deriv_xvector_test,
+    &deriv_S_and_b_test, &tot_objf_test, &tot_weight_test);
+
+  CuVector<BaseFloat> deriv_xvector_vec(xvector_dim);
+  deriv_xvector_vec.AddRowSumMat(1.0, deriv_xvector, 0.0);
+  CuVector<BaseFloat> deriv_xvector_vec_test(xvector_dim);
+  deriv_xvector_vec_test.AddRowSumMat(1.0, deriv_xvector_test, 0.0);
+
+  // Verify that the objfs are the same.
+  KALDI_ASSERT(ApproxEqual(tot_objf, tot_objf_test, 0.001));
+  // Also verify that the gradients are the same.
+  for (int32 i = 0; i < deriv_xvector_vec.Dim(); i++)
+    KALDI_ASSERT(ApproxEqual(deriv_xvector_vec(i), deriv_xvector_vec_test(i), 0.001));
+  for (int32 i = 0; i < deriv_S_and_b.Dim(); i++)
+    KALDI_ASSERT(ApproxEqual(deriv_S_and_b(i), deriv_S_and_b_test(i), 0.001));
+  return true;
+}
+
+void TestComputeXvectorObjfAndDeriv(
+    const CuMatrixBase<BaseFloat> &xvector_pairs,
+    const CuSpMatrix<BaseFloat> &S,
+    BaseFloat b, CuMatrixBase<BaseFloat> *deriv_xvector,
+    CuVector<BaseFloat> *deriv_S_and_b, BaseFloat *tot_objf,
```

Comment:
Please make this two outputs, a CuVector<BaseFloat>*deriv_S and a BaseFloat *deriv_b.
I am going to give these separate output nodes in the nnet, for easier diagnostics and for easier control of their learning rates.